File: cmd/uast/mapping.go
package main

import (
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"os"
	"strings"

	forest "github.com/alexaandru/go-sitter-forest"
	sitter "github.com/alexaandru/go-tree-sitter-bare"
	"github.com/spf13/cobra"

	"github.com/Sumatoshi-tech/codefang/pkg/uast/pkg/mapping"
)

// coveragePercent is the multiplier to convert a coverage ratio to a percentage.
const coveragePercent = 100

// Sentinel errors for the mapping command.
var (
	ErrNodeTypesRequired = errors.New("--node-types is required for non-treesitter operations")
	ErrNoInputFiles      = errors.New("no input files provided")
	ErrNoRootNode        = errors.New("no root node found")
)

func mappingCmd() *cobra.Command {
	var nodeTypesPath, mappingPath, format, language, extensions string

	var coverage, generate, showTreeSitter bool

	cmd := &cobra.Command{
		Use:   "mapping",
		Short: "UAST mapping helpers: grammar analysis, classification, coverage",
		Long:  `Analyze node-types.json, classify nodes, compute mapping coverage, and show tree-sitter JSON structure.`,
		RunE: func(_ *cobra.Command, args []string) error {
			return runMappingHelper(
				nodeTypesPath, mappingPath, format, coverage, generate,
				showTreeSitter, language, extensions, args,
			)
		},
	}

	cmd.Flags().StringVar(&nodeTypesPath, "node-types", "", "Path to node-types.json (required for non-treesitter operations)")
	cmd.Flags().StringVar(&mappingPath, "mapping", "", "Path to mapping DSL file (optional)")
	cmd.Flags().StringVar(&format, "format", "text", "Output format: text or json")
	cmd.Flags().BoolVar(&coverage, "coverage", false, "Compute mapping coverage if mapping is provided")
	cmd.Flags().BoolVar(&generate, "generate", false, "Generate .uastmap DSL from node-types.json")
	cmd.Flags().BoolVar(&showTreeSitter, "show-treesitter", false, "Show original tree-sitter JSON structure for input files")
	cmd.Flags().StringVar(&language, "language", "", "Language for tree-sitter parsing (language name or grammar file path)")
	cmd.Flags().StringVar(&extensions, "extensions", "", "Comma-separated list of file extensions for language declaration")

	return cmd
}

//nolint:cyclop,gocognit,gocyclo,funlen // mapping helper dispatches multiple operation modes
func runMappingHelper(
	nodeTypesPath, mappingPath, format string,
	coverage, generate, showTreeSitter bool,
	language, extensions string, args []string,
) error { //nolint:whitespace // multi-line function signature requires blank line after opening brace
	if showTreeSitter {
		return showTreeSitterJSON(args, language)
	}

	// Only require node-types when not showing tree-sitter JSON.
	if nodeTypesPath == "" {
		return ErrNodeTypesRequired
	}

	jsonData, err := os.ReadFile(nodeTypesPath)
	if err != nil {
		return fmt.Errorf("failed to read node-types.json: %w", err)
	}

	nodes, err := mapping.ParseNodeTypes(jsonData)
	if err != nil {
		return fmt.Errorf("failed to parse node-types.json: %w", err)
	}

	nodes = mapping.ApplyHeuristicClassification(nodes)

	if generate {
		// Parse extensions string into slice.
		var extensionsSlice []string

		if extensions != "" {
			extensionsSlice = strings.Split(extensions, ",")

			// Trim spaces from each extension.
			for idx, ext := range extensionsSlice {
				extensionsSlice[idx] = strings.TrimSpace(ext)
			}
		}

		dsl := mapping.GenerateMappingDSL(nodes, language, extensionsSlice)
		fmt.Print(dsl) //nolint:forbidigo // CLI user output

		return nil
	}

	var rules []mapping.MappingRule

	if mappingPath != "" {
		data, openErr := os.Open(mappingPath)
		if openErr != nil {
			return fmt.Errorf("failed to open mapping DSL: %w", openErr)
		}
		defer data.Close()

		_, _, parseErr := (&mapping.MappingParser{}).ParseMapping(data)
		if parseErr != nil {
			return fmt.Errorf("failed to load mapping DSL: %w", parseErr)
		}
	}

	if format == formatJSON { //nolint:nestif // JSON output with optional coverage requires nested conditionals
		out := map[string]any{
			"node_count": len(nodes),
			"categories": summarizeCategories(nodes),
			"nodes":      nodes,
		}

		if coverage && len(rules) > 0 {
			covResult, covErr := mapping.CoverageAnalysis(rules, nodes)
			if covErr != nil {
				return covErr
			}

			out["coverage"] = covResult
		}

		enc := json.NewEncoder(os.Stdout)
		enc.SetIndent("", "  ")

		encodeErr := enc.Encode(out)
		if encodeErr != nil {
			return fmt.Errorf("failed to encode JSON: %w", encodeErr)
		}

		return nil
	}

	fmt.Printf("Node types: %d\n", len(nodes)) //nolint:forbidigo // CLI user output

	cats := summarizeCategories(nodes)

	for cat, count := range cats {
		fmt.Printf("  %s: %d\n", cat, count) //nolint:forbidigo // CLI user output
	}

	if coverage && len(rules) > 0 {
		covResult, covErr := mapping.CoverageAnalysis(rules, nodes)
		if covErr != nil {
			return covErr
		}

		fmt.Printf("Coverage: %.2f%%\n", covResult*coveragePercent) //nolint:forbidigo // CLI user output
	}

	return nil
}

func showTreeSitterJSON(args []string, language string) error {
	if len(args) == 0 {
		return ErrNoInputFiles
	}

	for _, filename := range args {
		processErr := processFileForTreeSitterJSON(filename, language)
		if processErr != nil {
			return fmt.Errorf("failed to process %s: %w", filename, processErr)
		}
	}

	return nil
}

func processFileForTreeSitterJSON(filename, language string) error {
	content, err := os.ReadFile(filename)
	if err != nil {
		return fmt.Errorf("failed to read file: %w", err)
	}

	// Create a parser.
	parser := sitter.NewParser()

	// Set language if provided.
	if language != "" {
		lang := forest.GetLanguage(language)
		parser.SetLanguage(lang)
	}

	// Try to parse.
	tree, parseErr := parser.ParseString(context.Background(), nil, content)
	if parseErr != nil {
		if language == "" {
			return fmt.Errorf(
				"tree-sitter parsing requires a language to be set. Error: %w\n\n"+
					"Use --language flag to specify a language name or grammar file path", parseErr,
			)
		}

		return fmt.Errorf("failed to parse with tree-sitter: %w", parseErr)
	}

	root := tree.RootNode()
	if root.IsNull() {
		return ErrNoRootNode
	}

	jsonTree := convertTreeSitterNodeToJSON(root, content)

	fmt.Printf("=== Tree-sitter JSON for %s (language: %s) ===\n", filename, language) //nolint:forbidigo // CLI user output

	enc := json.NewEncoder(os.Stdout)
	enc.SetIndent("", "  ")

	encodeErr := enc.Encode(jsonTree)
	if encodeErr != nil {
		return fmt.Errorf("failed to encode JSON: %w", encodeErr)
	}

	fmt.Println() //nolint:forbidigo // CLI user output

	return nil
}

func convertTreeSitterNodeToJSON(tsNode sitter.Node, source []byte) map[string]any {
	result := map[string]any{
		"type": tsNode.Type(),
		"start_pos": map[string]int{
			"row":    int(tsNode.StartPoint().Row),    //nolint:gosec // tree-sitter coordinates fit in int
			"column": int(tsNode.StartPoint().Column), //nolint:gosec // tree-sitter coordinates fit in int
		},
		"end_pos": map[string]int{
			"row":    int(tsNode.EndPoint().Row),    //nolint:gosec // tree-sitter coordinates fit in int
			"column": int(tsNode.EndPoint().Column), //nolint:gosec // tree-sitter coordinates fit in int
		},
		"start_byte": int(tsNode.StartByte()), //nolint:gosec // tree-sitter byte offsets fit in int
		"end_byte":   int(tsNode.EndByte()),   //nolint:gosec // tree-sitter byte offsets fit in int
	}

	if tsNode.IsNamed() {
		result["named"] = true
	} else {
		result["named"] = false
	}

	// Extract text content.
	text := tsNode.Content(source)
	if text != "" {
		result["text"] = text
	}

	// Process children.
	childCount := tsNode.NamedChildCount()

	if childCount > 0 {
		children := make([]map[string]any, 0, childCount)

		for idx := range childCount {
			child := tsNode.NamedChild(idx)

			if !child.IsNull() {
				children = append(children, convertTreeSitterNodeToJSON(child, source))
			}
		}

		if len(children) > 0 {
			result["children"] = children
		}
	}

	return result
}

func summarizeCategories(nodes []mapping.NodeTypeInfo) map[string]int {
	cats := map[string]int{}

	for _, nodeInfo := range nodes {
		cats[fmt.Sprintf("%v", nodeInfo.Category)]++
	}

	return cats
}

File: rust/bins/uast/src/mapping.rs
//! `uast mapping` — UAST mapping helpers (port of `cmd/uast/mapping.go`).
//!
//! Analyzes `node-types.json`, classifies nodes, computes mapping coverage,
//! generates `.uastmap` DSL, or shows the raw tree-sitter JSON structure of an
//! input file. JSON output (`{node_count, categories, nodes[, coverage]}`) routes
//! through [`cf_textutil::write_json`] (DESIGN rule 1); `text` is human output.

use std::io::{self, Write};

use clap::{Arg, ArgAction, ArgMatches, Command};
use cf_textutil::{GoMap, GoValue};
use cf_uast_mapping::grammar_analysis::{
    apply_heuristic_classification, coverage_analysis, generate_mapping_dsl, parse_node_types,
};
use cf_uast_mapping::{NodeTypeInfo, Parser as MappingParser, Rule};

use crate::{COVERAGE_PERCENT, FORMAT_JSON};

/// Builds the `mapping` subcommand (mapping.go:38-69).
pub fn command() -> Command {
    Command::new("mapping")
        .about("UAST mapping helpers: grammar analysis, classification, coverage")
        .long_about(
            "Analyze node-types.json, classify nodes, compute mapping coverage, and show tree-sitter JSON structure.",
        )
        .arg(Arg::new("files").num_args(0..).index(1))
        .arg(long_opt("node-types", "", "Path to node-types.json (required for non-treesitter operations)"))
        .arg(long_opt("mapping", "", "Path to mapping DSL file (optional)"))
        .arg(long_opt("format", "text", "Output format: text or json"))
        .arg(long_flag("coverage", "Compute mapping coverage if mapping is provided"))
        .arg(long_flag("generate", "Generate .uastmap DSL from node-types.json"))
        .arg(long_flag("show-treesitter", "Show original tree-sitter JSON structure for input files"))
        .arg(long_opt("language", "", "Language for tree-sitter parsing (language name or grammar file path)"))
        .arg(long_opt("extensions", "", "Comma-separated list of file extensions for language declaration"))
}

/// Runs `mapping` (mapping.go `runMappingHelper`).
pub fn run(m: &ArgMatches) -> Result<(), String> {
    let node_types_path = m.get_one::<String>("node-types").map(String::as_str).unwrap_or("");
    let mapping_path = m.get_one::<String>("mapping").map(String::as_str).unwrap_or("");
    let format = m.get_one::<String>("format").map(String::as_str).unwrap_or("text");
    let coverage = m.get_flag("coverage");
    let generate = m.get_flag("generate");
    let show_treesitter = m.get_flag("show-treesitter");
    let language = m.get_one::<String>("language").map(String::as_str).unwrap_or("");
    let extensions = m.get_one::<String>("extensions").map(String::as_str).unwrap_or("");
    let files: Vec<String> =
        m.get_many::<String>("files").map(|v| v.cloned().collect()).unwrap_or_default();

    if show_treesitter {
        return show_tree_sitter(&files, language);
    }

    let nodes = load_node_types(node_types_path)?;

    if generate {
        let exts: Vec<String> = if extensions.is_empty() {
            Vec::new()
        } else {
            extensions.split(',').map(|s| s.trim().to_string()).collect()
        };
        let dsl = generate_mapping_dsl(&nodes, language, &exts);
        print!("{dsl}");
        return Ok(());
    }

    let rules = load_mapping_rules(mapping_path)?;

    if format == FORMAT_JSON {
        return output_json(&nodes, &rules, coverage);
    }
    output_text(&nodes, &rules, coverage)
}

/// Loads + classifies `node-types.json` (mapping.go `loadNodeTypes`). Errors
/// `--node-types is required for non-treesitter operations` when the path is
/// empty.
fn load_node_types(path: &str) -> Result<Vec<NodeTypeInfo>, String> {
    if path.is_empty() {
        return Err("--node-types is required for non-treesitter operations".to_string());
    }
    let data = std::fs::read(path).map_err(|e| format!("failed to read node-types.json: {e}"))?;
    let nodes = parse_node_types(&data).map_err(|e| format!("failed to parse node-types.json: {e}"))?;
    Ok(apply_heuristic_classification(nodes))
}

/// Loads mapping rules from a DSL file (mapping.go `loadMappingRules`). The Go
/// code parses the DSL only for validation and returns an empty rule list, which
/// is reproduced here.
fn load_mapping_rules(path: &str) -> Result<Vec<Rule>, String> {
    if path.is_empty() {
        return Ok(Vec::new());
    }
    let data = std::fs::read_to_string(path)
        .map_err(|e| format!("failed to open mapping DSL: {e}"))?;
    MappingParser::new()
        .parse_mapping(&data)
        .map_err(|e| format!("failed to load mapping DSL: {e}"))?;
    Ok(Vec::new())
}

/// Outputs the JSON summary (mapping.go `outputMappingJSON`).
fn output_json(nodes: &[NodeTypeInfo], rules: &[Rule], coverage: bool) -> Result<(), String> {
    let mut entries: Vec<(String, GoValue)> = vec![
        ("node_count".to_string(), GoValue::Int(nodes.len() as i64)),
        ("categories".to_string(), categories_value(nodes)),
        ("nodes".to_string(), nodes_value(nodes)),
    ];

    if coverage && !rules.is_empty() {
        let cov = coverage_analysis(rules, nodes).map_err(|e| e.to_string())?;
        entries.push(("coverage".to_string(), GoValue::Float(cov)));
    }

    // Go builds `map[string]any{...}` (map-origin, byte-sorted keys).
    let value = GoValue::object(GoMap::from_map(entries));
    let mut out = io::stdout();
    cf_textutil::write_json(&mut out, &value, true).map_err(|e| e.to_string())
}

/// Outputs the text summary (mapping.go `outputMappingText`).
fn output_text(nodes: &[NodeTypeInfo], rules: &[Rule], coverage: bool) -> Result<(), String> {
    let out = io::stdout();
    let mut out = out.lock();
    let _ = writeln!(out, "Node types: {}", nodes.len());
    for (cat, count) in summarize_categories(nodes) {
        let _ = writeln!(out, "  {cat}: {count}");
    }
    if coverage && !rules.is_empty() {
        let cov = coverage_analysis(rules, nodes).map_err(|e| e.to_string())?;
        let _ = writeln!(out, "Coverage: {:.2}%", cov * COVERAGE_PERCENT);
    }
    Ok(())
}

/// Summarizes nodes by category (mapping.go `summarizeCategories`).
///
/// Go does `cats[fmt.Sprintf("%v", nodeInfo.Category)]++`, and `NodeCategory` is
/// a bare `int` with no `String()` method, so `%v` renders the integer value
/// (`0`/`1`/`2`). The category key is therefore the integer rendered as a string.
fn summarize_categories(nodes: &[NodeTypeInfo]) -> Vec<(String, i64)> {
    use std::collections::BTreeMap;
    let mut cats: BTreeMap<String, i64> = BTreeMap::new();
    for n in nodes {
        *cats.entry(category_key(n.category)).or_insert(0) += 1;
    }
    cats.into_iter().collect()
}

/// Renders a [`NodeCategory`] as Go's `fmt.Sprintf("%v", category)` would: the
/// underlying `int` value (`Leaf`=0, `Container`=1, `Composite/Operator`=2).
fn category_key(c: cf_uast_mapping::NodeCategory) -> String {
    (c as i64).to_string()
}

/// Builds the `categories` map value (map-origin, byte-sorted).
fn categories_value(nodes: &[NodeTypeInfo]) -> GoValue {
    GoValue::object(GoMap::from_map(
        summarize_categories(nodes)
            .into_iter()
            .map(|(k, v)| (k, GoValue::Int(v)))
            .collect::<Vec<_>>(),
    ))
}

/// Builds the `nodes` array value: one struct-origin object per [`NodeTypeInfo`]
/// (mapping_types.go JSON tags: `name`, `children` omitempty, `fields` omitempty,
/// `category` as the integer iota value, `is_named`). Field declaration order is
/// preserved (struct-origin); `children`/`fields` are omitted when empty.
fn nodes_value(nodes: &[NodeTypeInfo]) -> GoValue {
    GoValue::Array(
        nodes
            .iter()
            .map(|n| {
                let mut entries = GoMap::new_struct();
                entries.push("name", GoValue::Str(n.name.clone()));
                if !n.children.is_empty() {
                    let kids = n
                        .children
                        .iter()
                        .map(|c| {
                            let mut ce = GoMap::new_struct();
                            ce.push("type", GoValue::Str(c.r#type.clone()));
                            ce.push("named", GoValue::Bool(c.named));
                            GoValue::Object(ce)
                        })
                        .collect();
                    entries.push("children", GoValue::Array(kids));
                }
                if !n.fields.is_empty() {
                    // fields is a map[string]FieldInfo (map-origin, byte-sorted).
                    let field_entries: Vec<(String, GoValue)> = n
                        .fields
                        .iter()
                        .map(|(k, fi)| {
                            let mut fe = GoMap::new_struct();
                            fe.push("name", GoValue::Str(fi.name.clone()));
                            fe.push(
                                "types",
                                GoValue::Array(
                                    fi.types.iter().cloned().map(GoValue::Str).collect(),
                                ),
                            );
                            fe.push("required", GoValue::Bool(fi.required));
                            fe.push("multiple", GoValue::Bool(fi.multiple));
                            (k.clone(), GoValue::Object(fe))
                        })
                        .collect();
                    entries.push("fields", GoValue::object(GoMap::from_map(field_entries)));
                }
                entries.push("category", GoValue::Int(n.category as i64));
                entries.push("is_named", GoValue::Bool(n.is_named));
                GoValue::Object(entries)
            })
            .collect(),
    )
}

/// Shows the raw tree-sitter JSON for input files (mapping.go
/// `showTreeSitterJSON`). The grammar wiring lives in `cf-uast` (DESIGN §5); this
/// reproduces the control flow and sentinel errors. Until grammars are vendored
/// the parse step is unavailable, so an unset/unknown language yields the same
/// `unsupported language` sentinel.
fn show_tree_sitter(files: &[String], language: &str) -> Result<(), String> {
    if files.is_empty() {
        return Err("no input files provided".to_string());
    }
    for filename in files {
        let _ = std::fs::read(filename)
            .map_err(|e| format!("failed to process {filename}: failed to read file: {e}"))?;
        // Language resolution mirrors mapping.go: an unknown/empty language is an
        // `unsupported language` error (the grammar dispatch is owned by cf-uast
        // and not yet wired; see cf-uast todos / DESIGN §5).
        if language.is_empty() {
            return Err(format!(
                "failed to process {filename}: tree-sitter parsing requires a language to be set"
            ));
        }
        return Err(format!("failed to process {filename}: unsupported language: {language}"));
    }
    Ok(())
}

fn long_opt(name: &'static str, default: &'static str, help: &'static str) -> Arg {
    Arg::new(name).long(name).help(help).default_value(default).action(ArgAction::Set)
}

fn long_flag(name: &'static str, help: &'static str) -> Arg {
    Arg::new(name).long(name).help(help).action(ArgAction::SetTrue)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn node_types_required_error() {
        let err = load_node_types("").unwrap_err();
        assert_eq!(err, "--node-types is required for non-treesitter operations");
    }

    #[test]
    fn show_treesitter_no_files_error() {
        let err = show_tree_sitter(&[], "go").unwrap_err();
        assert_eq!(err, "no input files provided");
    }

    #[test]
    fn empty_mapping_path_is_empty_rules() {
        assert!(load_mapping_rules("").unwrap().is_empty());
    }
}
